generic3g/specs/FieldSpec.F90

```diff
@@ -168,40 +176,96 @@ subroutine allocate(this, rc)
+         else if (this%vertical_dim == VERTICAL_DIM_EDGE) then
+            final_lbounds(1)=0
+            final_lbounds(2:)=this%ungridded_dims%get_lbounds()
+            final_ubounds(1)=num_levels
+            final_ubounds(2:)=this%ungridded_dims%get_ubounds()
+         end if
+      end if
+       
       call ESMF_FieldGet(this%payload, status=fstatus, _RC)
       if (fstatus == ESMF_FIELDSTATUS_GRIDSET) then
 
          call ESMF_FieldEmptyComplete(this%payload, this%typekind, &
-              ungriddedLBound= this%ungridded_dims%get_lbounds(),  &
-              ungriddedUBound= this%ungridded_dims%get_ubounds(),  &
+              ungriddedLBound= final_lbounds,  &
+              ungriddedUBound= final_ubounds,  &
               _RC)
          call ESMF_FieldGet(this%payload, status=fstatus, _RC)
          _ASSERT(fstatus == ESMF_FIELDSTATUS_COMPLETE, 'ESMF field status problem.')
 
          if (allocated(this%default_value)) then
-            if (this%typekind == ESMF_TYPEKIND_R4) then
-               block
-                 real(kind=ESMF_KIND_R4), pointer :: x(:,:)
-                 call ESMF_FieldGet(this%payload, farrayptr=x, _RC)
-                 x = this%default_value
-               end block
-            elseif (this%typekind == ESMF_TYPEKIND_R8) then
-               block
-                 real(kind=ESMF_KIND_R8), pointer :: x(:,:)
-                 call ESMF_FieldGet(this%payload, farrayptr=x, _RC)
-                 x = this%default_value
-               end block
-            else
-               _FAIL('unsupported typekind')
-            end if
+            call set_field_default(_RC)
          end if
           
          
          call this%set_allocated()
       end if
 
       _RETURN(ESMF_SUCCESS)
+
+      contains
+         subroutine set_field_default(rc)
+            integer, intent(out), optional :: rc
+            real(kind=ESMF_KIND_R4), pointer :: x_r4_1d(:),x_r4_2d(:,:),x_r4_3d(:,:,:),x_r4_4d(:,:,:,:)
+            real(kind=ESMF_KIND_R8), pointer :: x_r8_1d(:),x_r8_2d(:,:),x_r8_3d(:,:,:),x_r8_4d(:,:,:,:)
+            integer :: status, rank
+                
+            call ESMF_FieldGet(this%payload,rank=rank,_RC) 
+            if (this%typekind == ESMF_TYPEKIND_R4) then
+               if (rank == 1) then
+                  call ESMF_FieldGet(this%payload,farrayptr=x_r4_1d,_RC)
+                  x_r4_1d = this%default_value   
+               else if (rank == 2) then
+                  call ESMF_FieldGet(this%payload,farrayptr=x_r4_2d,_RC)
+                  x_r4_2d = this%default_value   
+               else if (rank == 3) then
+                  call ESMF_FieldGet(this%payload,farrayptr=x_r4_3d,_RC)
+                  x_r4_3d = this%default_value   
+               else if (rank == 4) then
+                  call ESMF_FieldGet(this%payload,farrayptr=x_r4_4d,_RC)
+                  x_r4_4d = this%default_value   
+               else
+                  _FAIL('unsupported rank')
```

Comment:
Can this conditional be pushed down into "Darian's" `geom` layer?   I'm really hoping to avoid conditional logic on rank here.   TYPEKIND as well, but that is harder.

Reply:
ok